### PR TITLE
Fix memory planning issues

### PR DESCRIPTION
**Pull request by pengwa**
**Description**: Few Regressions on Memory Planning

This is to fix few bugscaused by commit (https://github.com/microsoft/onnxruntime/commit/9af0d485241d3f037b94ebf876e457f396b6e583):
 
**Motivation and Context**
- Why is this change required? What problem does it solve?

1. Some of the initializers (that are in initializer_allocation_order ) are allocated more than once. 

2. TryResolveShape bug: not consider the resolved_shape is {}, which should be a valid shape, but actually considered as a failed resolving. This + "3" make ZeRO enabled cases have most of the activations not in planning, so they are changed to dynamic allocations, which explain what Wei have found in her traces.

3. For activation planner, if one of the shapes for inputs of NcclAllReduce/NcclAllGather/NcclReduceScatter is missing, then all activation planning is interrupted (and logged failed status "Unknown shape found in memory pattern compute"). While Previously, if one of the shape is missing for activation planner, we will skip that one, and continue planning for other activations. I think we might consider to continue instead of return a FAIL status. ( I did not change this in this PR, instead, I fixed 2, so there are no more shapes missing, then we won't return here .)

Another minor comment is: seems we are checking whether we need allocate contiguous memory by kernel definition, for example, NcclAllReduce/NcclAllGather/NcclReduceScatter, in some cases, actually the inputs could be non-contiguous for example following case. The value to reduce is a scalar. Making it contiguous won't bring something bad, but it would require all the shapes be there until behavior 3 is changed. 

- If it fixes an open issue, please link to the issue here.

**Validation**

https://github.com/microsoft/onnxruntime/commit/d98062da0ce1ffb804325396cde10a41933cd154: the commit before the commit https://github.com/microsoft/onnxruntime/commit/9af0d485241d3f037b94ebf876e457f396b6e583. 

https://github.com/microsoft/onnxruntime/commit/2ba637c55836c7b2dca554d96bf993970d086606: latest master

PR: current PR synced with the above lastest master commit. 

image

You will see:
- The memory usage (Green numbers) have some issues on latest master for ZeRO-enabled cases. Initializers are bigger, activation memory cannot planned due to TryResolveShape issue.
- With fix in this PR, all numbers are back to normal. While we did see some memory saving (dozens of MB) on activations for batch size 1. Remember this is a fairly small model, if we have more layers, the saving could be more.  

Add validation result for BERT-LARGE and GPT2-MEDIUM running from C++ frontend

image

### CODE REVIEW

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 1]
```diff
@@ -504,32 +504,30 @@ void RegisterTrainingOpSchemas() {
...
-           for (size_t contrib_node_index : {2, 4, 10}) {
-             nodes[contrib_node_index].set_domain(kMSDomain);
-           }
-           return nodes;
+            {// nodes: {outputs, op, inputs, attributes}
+
+             // Get input shapes and dynamic reduction axes.
+             {{"shape_A"}, "Shape", {"A"}},
+             {{"shape_B"}, "Shape", {"B"}},
+             {{"axes_A", "axes_B"}, "BroadcastGradientArgs", {"shape_A", "shape_B"}},
```
Comment (codemzs) [inline orttraining/orttraining/core/graph/training_op_defs.cc:512]: the changes in this file are not described in your PR, is this fixing another issue?

**Review by codemzs — APPROVED** [review]
:shipit:

**Review by codemzs — DISMISSED** [review]

**Review by wezuo — APPROVED** [review]

**Review by wezuo — COMMENTED** [review]
Could you check whether this would cause regression on existing models like bert-L and GPT?

**Review by wezuo — APPROVED** [review]

**Comment by pengwa** [comment]
> Could you check whether this would cause regression on existing models like bert-L and GPT before merge?

Done, please check the above results.

**Pull request merged**